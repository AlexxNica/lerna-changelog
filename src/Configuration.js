import LernaRepo from "lerna/lib/Repository";
import ConfigurationError from "./ConfigurationError";

export function fromCWD() {
  const lerna = new LernaRepo();

  const config = lerna.lernaJson.changelog;

  if (!config) {
    throw new ConfigurationError(
      "Missing changelog config in `lerna.json`.\n" +
      "See docs for setup: https://github.com/lerna/lerna-changelog#readme"
    );
  }

  config.rootPath = lerna.rootPath;

  return config;
}
