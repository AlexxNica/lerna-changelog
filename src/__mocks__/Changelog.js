const Changelog = require.requireActual("../Changelog").default;

const defaultConfig = {
  rootPath: "../",
  repo: "lerna/lerna-changelog",
  labels: {
    "Type: New Feature": ":rocket: New Feature",
    "Type: Breaking Change": ":boom: Breaking Change",
    "Type: Bug": ":bug: Bug Fix",
    "Type: Enhancement": ":nail_care: Enhancement",
    "Type: Documentation": ":memo: Documentation",
    "Type: Maintenance": ":house: Maintenance"
  },
  cacheDir: ".changelog"
};
const defaultListOfUniquePackages = [ "pkg-1", "pkg-2" ];
const defaultListOfCommits = [
  "a0000005;HEAD -> master, tag: v0.2.0, origin/master, origin/HEAD;chore(release): releasing component;2017-01-01",
  "a0000004;;Merge pull request #2 from my-feature;2017-01-01",
  "a0000003;;feat(module) Add new module (#2);2017-01-01",
  "a0000002;;refactor(module) Simplify implementation;2017-01-01",
  "a0000001;tag: v0.1.0;chore(release): releasing component;2017-01-01"
];

let currentConfig = defaultConfig;
let currentListOfCommits = defaultListOfCommits;

export function __resetDefaults() {
  currentConfig = defaultConfig;
  currentListOfCommits = defaultListOfCommits;
}

export function __getConfig() {
  return currentConfig;
}
export function __setConfig(customConfig) {
  currentConfig = { ...defaultConfig, ...customConfig };
}
export function __prependListOfCommits(customListOfCommits) {
  currentListOfCommits = [ ...customListOfCommits, ...defaultListOfCommits ];
}

class MockedChangelog extends Changelog {
  getConfig() {
    return currentConfig;
  }
  getListOfUniquePackages() {
    return defaultListOfUniquePackages;
  }
  getListOfCommits() {
    return currentListOfCommits;
  }
}

export default MockedChangelog;
