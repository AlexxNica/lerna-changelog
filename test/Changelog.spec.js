jest.mock("lerna/lib/Repository");
jest.mock("lerna/lib/progressBar");
jest.mock("../src/ApiDataCache");
jest.mock("../src/ConfigurationError");
jest.mock("../src/GithubAPI");
jest.mock("../src/Changelog");

describe("contructor", () => {
  const MockedChangelog = require("../src/Changelog").default;
  const testConfig = require("../src/Changelog").__getConfig();

  beforeEach(() => {
    require("../src/Changelog").__resetDefaults();
  })

  it("set config", () => {
    const changelog = new MockedChangelog();
    expect(changelog.config).toEqual(testConfig);
  });
  it("set remote", () => {
    const changelog = new MockedChangelog();
    expect(changelog.remote).toBeDefined();
  });
  it("set cli options", () => {
    const changelog = new MockedChangelog({ "tag-from": "1", "tag-to": "2" });
    expect(changelog.tagFrom).toBe("1");
    expect(changelog.tagTo).toBe("2");
  });
});

describe("getCommitsInfo", () => {
  const MockedChangelog = require("../src/Changelog").default;
  const changelog = new MockedChangelog();
  const commitsInfo = changelog.getCommitsInfo();

  it("parse commits with different tags", () => {
    expect(commitsInfo).toEqual([
      {
        commitSHA: "a0000005",
        date: "2017-01-01",
        labels: [],
        message: "chore(release): releasing component",
        tag: "v0.2.0"
      },
      {
        commitSHA: "a0000004",
        date: "2017-01-01",
        labels: [
          { name: "Type: New Feature" },
          { name: "Status: In Progress" },
        ],
        mergeMessage: "Merge pull request #2 from my-feature",
        message: "Merge pull request #2 from my-feature",
        tag: "",
        title: "This is the commit title for the issue (#2)",
        user: {
          html_url: "https://github.com/test-user",
          login: "test-user",
          name: "Test User"
        }
      },
      {
        commitSHA: "a0000003",
        date: "2017-01-01",
        labels: [
          { name: "Type: New Feature" },
          { name: "Status: In Progress" },
        ],
        mergeMessage: "feat(module) Add new module (#2)",
        message: "feat(module) Add new module (#2)",
        tag: "",
        title: "This is the commit title for the issue (#2)",
        user: {
          html_url: "https://github.com/test-user",
          login: "test-user",
          name: "Test User"
        }
      },
      {
        commitSHA: "a0000002",
        date: "2017-01-01",
        labels: [],
        message: "refactor(module) Simplify implementation",
        tag: ""
      },
      {
        commitSHA: "a0000001",
        date: "2017-01-01",
        labels: [],
        message: "chore(release): releasing component",
        tag: "v0.1.0"
      }
    ]);
  });
});

describe("getCommitsByCategory", () => {
  const MockedChangelog = require("../src/Changelog").default;
  const changelog = new MockedChangelog();
  const testCommits = [
    { commitSHA: "a0000005", labels: [{ name: "Status: In Progress" }] },
    { commitSHA: "a0000004", labels: [{ name: "Type: Bug" }] },
    {
      commitSHA: "a0000003",
      labels: [
        { name: "Type: New Feature" },
        { name: "Status: In Progress" },
      ]
    },
    { commitSHA: "a0000002", labels: [] },
    { commitSHA: "a0000001", labels: [] }
  ]
  const commitsByCategory = changelog.getCommitsByCategory(testCommits);

  it("group commits by category", () => {
    expect(commitsByCategory).toEqual([
      {
        commits: [
          {
            commitSHA: "a0000003",
            labels: [
              { name: "Type: New Feature" },
              { name: "Status: In Progress" }
            ]
          }
        ],
        heading: ":rocket: New Feature"
      },
      { commits: [], heading: ":boom: Breaking Change" },
      {
        commits: [
          { commitSHA: "a0000004", labels: [{ name: "Type: Bug" }] }
        ],
        heading: ":bug: Bug Fix"
      },
      { commits: [], heading: ":nail_care: Enhancement" },
      { commits: [], heading: ":memo: Documentation" },
      { commits: [], heading: ":house: Maintenance" }
    ]);
  })
})

describe("getCommitters", () => {
  require("../src/Changelog").__setConfig({ ignoreCommitters: ["user-bot"] });
  const MockedChangelog = require("../src/Changelog").default;
  const changelog = new MockedChangelog();

  const testCommits = [
    { commitSHA: "a0000004", user: { login: "test-user-1" } },
    { commitSHA: "a0000003", user: { login: "test-user-2" } },
    { commitSHA: "a0000002", user: { login: "user-bot" } },
    { commitSHA: "a0000001" },
  ]
  const committers = changelog.getCommitters(testCommits);

  it("get list of valid commiters", () => {
    expect(committers).toEqual([
      "Test User ([test-user-1](https://github.com/test-user))",
      "Test User ([test-user-2](https://github.com/test-user))"
    ]);
  })
})

describe("createMarkdown", () => {
  const MockedChangelog = require("../src/Changelog").default;
  const changelog = new MockedChangelog();

  beforeEach(() => {
    require("../src/GithubAPI").__resetDefaults();
    require("../src/Changelog").__resetDefaults();
  })

  it("get markdown grouped by tags", () => {
    require("../src/Changelog").__prependListOfCommits([
      // Add some commits that do not belong to a tag yet
      "a0000008;;Merge pull request #3 from my-fix-feature;2017-01-02",
      "a0000007;;feat(module): some unreleased feature;2017-01-02",
      "a0000006;;fix(module): fix a problem (#2);2017-01-02",
    ]);
    const markdown = changelog.createMarkdown();
    expect(markdown).toMatchSnapshot();
  })

  it("get markdown grouped by tags (with commit number link)", () => {
    const issue = require("../src/GithubAPI").createTestIssue(1);
    require("../src/GithubAPI").__setIssue({ ...issue, number: 1 });
    const markdown = changelog.createMarkdown();
    expect(markdown).toMatchSnapshot();
  })

  it("get markdown grouped by tags (with matching fix commit)", () => {
    const issue = require("../src/GithubAPI").createTestIssue(1);
    require("../src/GithubAPI").__setIssue({
      ...issue,
      title: "refactor(module): something. Closes #1"
    });
    const markdown = changelog.createMarkdown();
    expect(markdown).toMatchSnapshot();
  })
})
