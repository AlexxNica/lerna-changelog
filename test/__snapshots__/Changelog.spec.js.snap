exports[`createMarkdown get markdown grouped by tags (with commit number link) 1`] = `
"
## v0.2.0 (2017-01-01)

#### :rocket: New Feature
* \`pkg-1\`, \`pkg-2\`
  * [#1](https://github.com/lerna/lerna-changelog/pull/1) This is the commit title for the issue (#1). ([@test-user](https://github.com/test-user))
  * [#1](https://github.com/lerna/lerna-changelog/pull/1) This is the commit title for the issue (#1). ([@test-user](https://github.com/test-user))

#### Committers: 1
- Test User ([test-user](https://github.com/test-user))"
`;

exports[`createMarkdown get markdown grouped by tags (with matching fix commit) 1`] = `
"
## v0.2.0 (2017-01-01)

#### :rocket: New Feature
* \`pkg-1\`, \`pkg-2\`
  * refactor(module): something. Closes [#1](https://github.com/lerna/lerna-changelog/issues/1). ([@test-user](https://github.com/test-user))
  * refactor(module): something. Closes [#1](https://github.com/lerna/lerna-changelog/issues/1). ([@test-user](https://github.com/test-user))

#### Committers: 1
- Test User ([test-user](https://github.com/test-user))"
`;

exports[`createMarkdown get markdown grouped by tags 1`] = `
"
## Unreleased (2099-01-01)

#### :rocket: New Feature
* \`pkg-1\`, \`pkg-2\`
  * This is the commit title for the issue (#3). ([@test-user](https://github.com/test-user))
  * This is the commit title for the issue (#2). ([@test-user](https://github.com/test-user))

#### Committers: 1
- Test User ([test-user](https://github.com/test-user))


## v0.2.0 (2017-01-01)

#### :rocket: New Feature
* \`pkg-1\`, \`pkg-2\`
  * This is the commit title for the issue (#2). ([@test-user](https://github.com/test-user))
  * This is the commit title for the issue (#2). ([@test-user](https://github.com/test-user))

#### Committers: 1
- Test User ([test-user](https://github.com/test-user))"
`;
